### PR TITLE
torch.compile each TransformerBlock instead of the whole model

### PR DESCRIPTION
**Pull request by wanchaol**
This way we could temporarily enable 2-D parallel compile, and it might make sense to do transformer block compile in the future with PP (which we'll see).

We should figure out the dynamic shape issue though

### CODE REVIEW

File: torchtitan/parallelisms/parallelize_llama.py [hunk 1]
```diff
@@ -227,6 +227,19 @@ def parallelize_llama(model, world_mesh, parallel_dims, job_config: JobConfig):
...
                     transformer_block, job_config.activation_checkpoint
                 )
+
+            if job_config.training.compile:
+                if (
+                    job_config.activation_checkpoint.mode == "selective"
+                    and job_config.activation_checkpoint.selective_ac_option == "op"
+                ):
+                    torch._dynamo.config._experimental_support_context_fn_in_torch_utils_checkpoint = (
+                        True
```
Comment (tianyu-l) [inline torchtitan/parallelisms/parallelize_llama.py]: don't need to set this flag in every layer?

File: torchtitan/parallelisms/parallelize_llama.py [hunk 2]
```diff
@@ -227,6 +227,19 @@ def parallelize_llama(model, world_mesh, parallel_dims, job_config: JobConfig):
...
+                    )
+                logger.info(f"Compiling TransformerBlock {layer_id} with torch.compile")
```
Comment (tianyu-l) [inline torchtitan/parallelisms/parallelize_llama.py]: Let's only log once outside since there could be many layers. Since this file is parallelize_llama.py, can we consider put this per layer compiling back to train.py, or in a separate file (say, compile.py) if it's too bulky.

Comment (wanchaol) [reply torchtitan/parallelisms/parallelize_llama.py]: It's kinda hard, we have to put compile here bc we want to have compile apply to each TransformerBlock after AC, but before FSDP wrapping, to avoid the dynamic shape issue, otherwise 2D compile would still not work

File: torchtitan/parallelisms/parallelize_llama.py (hunk 1 above)
Comment (wanchaol) [reply torchtitan/parallelisms/parallelize_llama.py]: separate the flag setting back to train.py and in put the actual compile to `parallelize_llama.py`, lmk if that looks better

File: train.py [hunk 3]
```diff
@@ -219,17 +219,16 @@ def loss_fn(pred, labels):
...
             job_config.activation_checkpoint.mode == "selective"
             and job_config.activation_checkpoint.selective_ac_option == "op"
         ):
+            # some flags for torch.compile enablement
             torch._dynamo.config._experimental_support_context_fn_in_torch_utils_checkpoint = (
                 True
             )
-        logger.info("Compiling model with torch.compile")
-        model = torch.compile(model)
+        logger.info("Compiling each TransformerBlock with torch.compile")
```
Comment (tianyu-l) [inline train.py]: nit: Since we put parallelization, ac, and compile all in parallelize_llama.py, it might make more sense to put this all in that file. This will simplify `train.py`.

Concretely, we can put this block to `parallelize_llama.py`, right before `logger.info("Applied FSDP to the model")` and change the wording from `"Compiling each ..."` to `"Compiled each ..."`

File: train.py [hunk 4]
```diff
@@ -219,17 +219,16 @@ def loss_fn(pred, labels):
 
     metric_logger = build_metric_logger(job_config)
 
-    # torch.compile model for improved performance
     if job_config.training.compile:
```
Comment (tianyu-l) [inline train.py]: nit: since we are moving compiling to parallelize_llama, let's add some comment on L201.

**Review by tianyu-l — APPROVED** [review]
lgtm! Some nit comments on how to organize things.

File: torchtitan/parallelisms/parallelize_llama.py [hunk 5]
```diff
@@ -213,20 +213,30 @@ def parallelize_llama(model, world_mesh, parallel_dims, job_config: JobConfig):
...
+    ac_mode = job_config.activation_checkpoint.mode
+    for layer_id, transformer_block in enumerate(model.layers):
+        if ac_mode in ("full", "selective"):
+            transformer_block = checkpoint_wrapper(
+                transformer_block, job_config.activation_checkpoint
+            )
+            logger.info(f"Applied {ac_mode} activation checkpointing to the model")
+
+        if job_config.training.compile:
+            # turn on per-transformer block compile after AC wrappnig and before FSDP
```
Comment (tianyu-l) [inline torchtitan/parallelisms/parallelize_llama.py]: typo: wrappnig -> wrapping

File: torchtitan/parallelisms/parallelize_llama.py [hunk 6]
```diff
@@ -213,20 +210,46 @@ def parallelize_llama(model, world_mesh, parallel_dims, job_config: JobConfig):
...
+        if (
+            enable_compile
+            and ac_config.mode == "selective"
+            and ac_config.selective_ac_option == "op"
+        ):
+            # some temp flags for torch.compile enablement + SAC
+            torch._dynamo.config._experimental_support_context_fn_in_torch_utils_checkpoint = (
+                True
+            )
+    if enable_compile:
```
Comment (awgu) [inline torchtitan/parallelisms/parallelize_llama.py:389]: nit: Can this error check be moved up to line 216?

**Comment by awgu** [comment]
Should we be able to close https://github.com/pytorch/torchtitan/issues/61 after this PR?

Also, do we need to run end-to-end numerics testing?

**Comment by wanchaol** [comment]
> Should we be able to close #61 after this PR?
> 
> Also, do we need to run end-to-end numerics testing?

@awgu yeah I think it should resolve that, I'll do some e2e benchmarking before landing, so this would likely take a while

**Comment by wanchaol** [comment]
I'll break up some other changes to land them first

**Comment by kwen2501** [comment]
when PP is present, we may torch.compile the whole stage module, which is bigger than a transformer block, i.e.

```
pipe = pipeline(model, ...)
stage_mod = pipe.get_stage_module(stage_idx)
stage_mod = torch.compile(stage_mod)
stage = PipelineStage(stage_mod, ...)
```

It would also allow the code to be more model-agnostic -- there is no `transformer_block`, `layer_id` or `model.layers` here.

**Comment by chrisociepa** [comment]
Apart from compiling each layer separately, you could also compile tok_embeddings, output and norm in the same way

**Comment by chrisociepa** [comment]
In my case, enabling compilation in this way (per-layer) causes a memory leak

**Comment by wanchaol** [comment]
> In my case, enabling compilation in this way (per-layer) causes a memory leak

🤔 interesting, how did you observe that?

fwiw this doesn't work out of box, as it trigger some non-trival numeric issues, I'm going to leave this PR here until I resolved it. Opening a new PR to turn dynamic shape off so that it works for both 1D and 2D compile

**Comment by chrisociepa** [comment]
With each iteration, the memory usage increases and eventually results in OOM. However, just to be clear, I haven't tested this on your entire code, only on a part of it. Adding per-layer compilation causes a memory leak with each iteration. I know that the memory leak might be related to my implementation, so I just wanted to bring this issue to your attention. If you don't observe this in your code, then it's likely an issue on my end.
Meanwhile, I didn't observe the numerical issues you mentioned, unless they are a direct consequence of the memory leak. The loss function appears practically the same with layer compilation both enabled and disabled.

File: torchtitan/parallelisms/parallelize_llama.py [hunk 7]
```diff
@@ -213,20 +210,46 @@ def parallelize_llama(model, world_mesh, parallel_dims, job_config: JobConfig):
...
+    # apply AC + torch.compile
+    ac_config = job_config.activation_checkpoint
+    enable_compile = job_config.training.compile
+    for layer_id, transformer_block in enumerate(model.layers):
+        if ac_config.mode in ("full", "selective"):
+            transformer_block = checkpoint_wrapper(transformer_block, ac_config)
+        if enable_compile:
+            # turn on per-transformer block compile after AC wrapping and before FSDP
+            # TODO: dynamic shape have some issues so we turn it off for now.
+            transformer_block = torch.compile(transformer_block, dynamic=False)
```
Comment (yifuwang) [inline torchtitan/parallelisms/parallelize_llama.py:375]: Curious how far we are from being able to enable `fullgraph=True`?

Comment (wanchaol) [reply torchtitan/parallelisms/parallelize_llama.py:375]: I think `fullgraph=True` should work already when we moving to per-TransformerBlock compile, maybe I can add that flag too.

I ended up just turn dynamic=False in the current full mode compile https://github.com/pytorch/torchtitan/pull/297 in this case we can't full_graph=True yet as FSDP still graph breaking, but for that case we should also captured each TransformerBlock as full graphs already

**Comment by wanchaol** [comment]
going to merge this given that:
1. 2D compile currently broken and this PR work arounds it and make TP can be compiled again (we should separately figure out the full model compile issue) cc @bdhirsh 
2. per-TransformerBlock compile would give us potential later once the cache reusing in torch.compile enabled, it would drastically improve the compile (code start and warm start) time.
